lib/route_consistency_checker.rb

```diff
@@ -0,0 +1,143 @@
+      .where(:schema_name.not => /^placeholder/)
+  end
+
+  def unchecked_routes
+    Set.new(routes.keys) - checked_routes
+  end
+
+  def get_route(path)
+    begin
+      route = routes.fetch(path.to_sym)
+      checked_routes.add?(path.to_sym)
+      route
+    rescue KeyError
+      errors[path] << "Path (#{path}) was not found!"
+      nil
+    end
+  end
+
+  def find_content_item_for_field(path, kind)
+    items = ContentItem.where("#{kind}s.path" => path).entries
+    return nil if items.empty?
+    return items.first if items.length == 1
+    errors[path] << "Multiple content items returned for #{kind}."
+    nil
+  end
+
+  def find_content_item_by_route(path)
+    find_content_item_for_field(path, "route")
+  end
+
+  def find_content_item_by_redirect(path)
+    find_content_item_for_field(path, "redirect")
+  end
+
+  def find_content_item(path)
+    # We cannot use an 'or' query here as it seems to be slower than doing this
+    # irb(main):019:0> time { ContentItem.where(routes: { "$elemMatch" => { path: path } }).first }
+    #   0.000000   0.000000   0.000000 (  0.006512)
+    # irb(main):020:0> time { ContentItem.where(redirects: { "$elemMatch" => { path: path } }).first }
+    #   0.010000   0.000000   0.010000 (  0.004688)
+    # irb(main):021:0> time { ContentItem.or({ redirects: { "$elemMatch" => { path: path } } }, { routes: { "$elemMatch" => { path: path } } }).first }
+    #   0.000000   0.000000   0.000000 ( 11.314430)
+    find_content_item_by_route(path) || find_content_item_by_redirect(path)
+  end
+
+  def check_redirect(content_item, redirect)
+    path = redirect[:path]
+
+    result = get_route(path)
+    return unless result
+
+    return if content_item.updated_at > result.updated_at
+
+    if result.handler != "redirect"
+      errors[path] << "Handler should be a redirect."
+    end
+
+    if result.redirect_to != redirect[:destination]
+      errors[path] << "Route destination (#{result.redirect_to}) does not " \
+                      "match item destination (#{redirect[:destination]})."
+    end
+  end
+
+  def check_route(content_item, route)
+    path = route[:path]
+
+    result = get_route(path)
+    return unless result
+
+    return if content_item.updated_at > result.updated_at
+
+    if result.handler != expected_handler(content_item)
+      errors[path] << "Handler (#{result.handler}) does not match expected " \
+                      "item handler (#{expected_handler(content_item)})."
+    end
+
+    if result.backend_id != content_item.rendering_app
+      errors[path] << "Backend ID (#{result.backend_id}) does not match " \
+                      "item rendering app (#{content_item.rendering_app})."
+    end
+
+    errors[path] << "Route is marked as disabled." if result.disabled
+  end
+
+  def expected_handler(content_item)
+    if content_item.gone?
+      "gone"
+    elsif content_item.redirect?
+      "redirect"
+    else
+      "backend"
+    end
```

Comment:
might be cleaner as
return "gone" if content_item.gone?
return "redirect" if content_item.redirect?
"backend"